core/src/main/java/org/apache/iceberg/BaseTableScan.java

```diff
@@ -86,20 +90,56 @@ public TableScan appendsAfter(long fromSnapshotId) {
   public TableScan useSnapshot(long scanSnapshotId) {
     Preconditions.checkArgument(
         snapshotId() == null, "Cannot override snapshot, already set to id=%s", snapshotId());
+    Preconditions.checkArgument(
+        ref() == null, "Cannot override snapshot, already set to id=%s", ref());
     Preconditions.checkArgument(
         tableOps().current().snapshot(scanSnapshotId) != null,
         "Cannot find snapshot with ID %s",
         scanSnapshotId);
     return newRefinedScan(
-        tableOps(), table(), tableSchema(), context().useSnapshotId(scanSnapshotId));
+        tableOps(),
+        table(),
+        tableSchema(),
+        context().useRef(SnapshotRef.MAIN_BRANCH).useSnapshotId(scanSnapshotId));
+  }
+
+  @Override
+  public TableScan useRef(String name) {
+    Preconditions.checkArgument(
+        snapshotId() == null, "Cannot override ref, already set to id=%s", snapshotId());
```

Comment:
`already set snapshot id=`?

Reply:
I was following the existing snapshot id already set error messaging, I'm definitely open to change it but then I think we should change it in the other places, let me know what you think!

Reply:
I think `set snapshot id=` is more clear, otherwise people might misunderstand and think there is a ref id.

Reply:
Good point, I agree. I've updated the error messages in all the places.